Doc updates for ChainRulesCore 0.5 (#152)
JuliaDiff/ChainRules.jl: docs/src/index.md

**ChainRules is an AD-independent collection of rules to use in a differentiation system.**

### Introduction

!!! note "The whole field is a mess for terminology"
    It isn't just ChainRules, it is everyone.
    Internally ChainRules tries to be consistent.
    Help with that is always welcomed.

!!! terminology "Primal"
Often we will talk about something as _primal_.
That means it is related to the original problem, not its derivative.
For example for `y = foo(x)`
`foo` is the _primal_ function,
computing `foo(x)` is doing the _primal_ computation.
`y` is the _primal_ return, and `x` is a _primal_ argument.
`typeof(y)` and `typeof(x)` are both _primal_ types.

### `frule` and `rrule`

!!! terminology "`frule` and `rrule`"
    `frule` and `rrule` are ChainRules specific terms.
    Their exact functioning is fairly ChainRules specific, though other tools have similar functions.
    The core notion is sometimes called _custom AD primitives_, _custom adjoints_, _custom_gradients_, _custom sensitivities_.

> The rules are encoded as `rrule`s and `frule`s, for use in forward-mode and reverse-mode differentiation respectively.

Reviewer: ```suggestion
The rules are encoded as `frule`s and `rrule`s, for use in forward-mode and reverse-mode differentiation respectively.
```

The `rrule` for some function `foo`, which takes the positional arguments `args` and keyword arguments `kwargs`, is written:

```julia
function rrule(::typeof(foo), args...; kwargs...)
    ...
    return y, pullback
end
```

> Where `y` (the primal result) must be equal to `foo(args...; kwargs...)`.

Reviewer: ```suggestion
where `y` (the primal result) must be equal to `foo(args...; kwargs...)`.
```

`pullback` is a function to propagate the derivative information backwards at that point.
That pullback function is used:
`∂self, ∂args... = pullback(Δy)`

> function takes arguments;:  (more later).

Author: I think I got distracted when writing it, it probably can be deleted.

> function takes arguments;:  (more later).

Reviewer 2: ```suggestion
```

Considered alone it would look like:

```julia
function pushforward(Δself, Δargs...)
    ...
    return ∂y
end
```
But because it is fused into frule we see it as part of:
```julia
function frule(::typeof(foo), args..., Δself, Δargs...; kwargs...)
    ...
    return y, ∂y
end
```

The input to the pushforward is often called the _perturbation_.
If the function is `y = f(x)` often the pushforward will be written `ẏ = last(frule(f, x, ṡelf, ẋ))`.
`ẏ` is commonly used to represent the perturbation for `y`.

!!! note
    
    In the `frule`/pushforward,
    there is one `Δarg` per `arg` to the original function.
    The `Δargs` are similar in type/structure to the corresponding inputs `args` (`Δself` is explained below).
    The `∂y` are similar in type/structure to the original function's output `Y`.
    In particular if that function returned a tuple then `∂y` will be a tuple of the same size.

### Self derivative `Δself`, `∂self`, `s̄elf`, `ṡelf` etc.

!!! terminology  "`Δself`, `∂self`, `s̄elf`, `ṡelf`"
    It is the derivatives with respect to the internal fields of the function.
    To the best of our knowledge there is no standard terminology for this.

>     Other good names might be `Δinternal`/`∂internal`.

Reviewer 2: hmmm, this still doesn't seem to render right... may something like
```julia
!!! terminology "self"
    `Δself`, `∂self`, `s̄elf`, `ṡelf` denote the derivatives with respect to the internal fields of the function.
    To the best of our knowledge there is no standard terminology for this.
    Other good names might be `Δinternal`/`∂internal`.
```

Author: I think just dropping the backticks is best

#### Pushforward / Pullback summary

- **Pullback**
   - returned by `rrule`
   - takes output space wobbles, gives input space wiggles
   - 1 argument per original function return
   - 1 return per original function argument + 1 for the function itself

- **Pushforward:**
    - part of `frule`
    - takes input space wiggles, gives output space wobbles
    - 1 argument per original function argument + 1 for the function itself
    - 1 return per original function return

#### Pullback/Pushforward and Directional Derivative/Gradient

The most trivial use of the `pushforward` from within `frule` is to calculate the directional derivative:

> If we would like to know the the directional derivative of `f` for an input chage of `(1.5, 0.4, -1)`

Reviewer 2: ```suggestion
If we would like to know the the directional derivative of `f` for an input change of `(1.5, 0.4, -1)`
```

```julia
direction = (1.5, 0.4, -1) # (ȧ, ḃ, ċ)

> y, ẏ = frule(f, a, b, c, Zero(), 1, 0, 0)

Reviewer: Should the derivative arguments here not be `direction`?

Author: They should. Good catch

```

On the basis directions one gets the partial derivatives of `y`:
```julia
y, dy_da = frule(f, a, b, c, Zero(), 1, 0, 0)
y, dy_db = frule(f, a, b, c, Zero(), 0, 1, 0)

> y, dy_dc = frule(f, a, b, c, Zero(), 0, 0, 1)

Reviewer 2: `dy_da` could also be written `ā`, and `dy_db` could be written `b̄` and so on?

Reviewer: The overbar notation is usually reserved for reverse-mode-related quantities, so probably best not to do that here. Might make sense to write these as partials though (to be pedantic):
```julia
∂y_∂a
```
etc

```

Similarly, the most trivial use of `rrule` and returned `pullback` is to calculate the [Gradient](https://en.wikipedia.org/wiki/Gradient):

```julia
y, f_pullback = rrule(f, a, b, c)
∇f = f_pullback(1)  # for appropriate `1`-like seed.
s̄elf, ā, b̄, c̄ = ∇f
```
Then we have that `∇f` is the _gradient_ of `f` at `(a, b, c)`.
And we thus have the partial derivatives ``\overline{\mathrm{self}}, = \dfrac{∂f}{∂\mathrm{self}}``, ``\overline{a} = \dfrac{∂f}{∂a}``, ``\overline{b} = \dfrac{∂f}{∂b}``, ``\overline{c} = \dfrac{∂f}{∂c}``, including the and the self-partial derivative, ``\overline{\mathrm{self}}``.

> ### Differentials

Reviewer: I think this whole section needs to be expanded on to make various points:

- `+` is the only operation guaranteed to be defined on differentials and, moreover, it's only guaranteed to be defined between differentials that are valid for the same primal type.
- `*` is only guaranteed to be defined between scalars and differentials. You can't generally multiple a differential by another differential
- `+` between primals and differentials isn't always defined, but it is for a lot of interesting cases e.g. `Real`, `Matrix{<:Real}` etc
- stuff that I've missed that we've figured out since this section was last written.

Author: Yep. But out of scope for this PR

The values that come back from pullbacks or pushforwards are not always the same type as the input/outputs of the primal function.
They are differentials, which correspond roughly to something able to represent the difference between two values of the primal types.
A differential might be such a regular type, like a `Number`, or a `Matrix`, matching to the original type;
or it might be one of the `AbstractDifferential` subtypes.

Differentials support a number of operations.
Most importantly: `+` and `*`, which let them act as mathematical objects.

The most important `AbstractDifferential`s when getting started are the ones about avoiding work:

>  - `Thunk`: this is a deferred computation. A thunk is a [word for a zero argument closure](https://en.wikipedia.org/wiki/Thunk). A computation wrapped in a `@thunk` doesn't get evaluated until `unthunk` is called on the `Thunk`. (`unthunk` is a no-op on nonthunked inputs.)

Reviewer 2: ```suggestion
 - `Thunk`: this is a deferred computation. A thunk is a [word for a zero argument closure](https://en.wikipedia.org/wiki/Thunk). A computation wrapped in a `@thunk` doesn't get evaluated until `unthunk` is called on the `Thunk`. `unthunk` is a no-op on inputs that are not `Thunk`s.
```

Author: Not quite since both `Thunk` and `InplaceThunk` are thunks

Reviewer 2: Ah I forgot Thunk isn't the abstract type. 

```suggestion
 - `Thunk`: this is a deferred computation. A thunk is a [word for a zero argument closure](https://en.wikipedia.org/wiki/Thunk). A computation wrapped in a `@thunk` doesn't get evaluated until `unthunk` is called on the `Thunk`. `unthunk` is a no-op on non-thunked inputs.
```

>  - `One`, `Zero`: There are special representations of `1` and `0`. They do great things around avoiding expanding `Thunks` in multiplication and (for `Zero`) addition.

Reviewer: ```suggestion
 - `Zero`: There is a special representation of `0`. It's defined as the differential that leaves and differential that it's added to unchanged. It's great for avoiding computation.
```
I just don't really want to talk about `One` in the docs since it's looking increasingly likely that we'll drop it in the near future.

Reviewer 2: i don't feel strongly, but i'd slightly leave change to `One` in docs to be included in (or after) any changes to `One` itself

Reviewer: Yeah, you're probably right. I'm being overzealous.

#### Other `AbstractDifferential`s:

>  - `Composite{P}`: this is the differential for tuples and  structs. Use it like a tuple or named tuple. The type parameter `P` is for the primal type.

Reviewer 2: ```suggestion
 - `Composite{P}`: this is the differential for tuples and  structs. Use it like a `Tuple` or `NamedTuple`. The type parameter `P` is for the primal type.
```

>  - `DoesNotExist`: Zero-like, represents that the operation on this input is not differentiable. Its primal type is normally integer or boolean.

Reviewer 2: ```suggestion
 - `DoesNotExist`: Zero-like, represents that the operation on this input is not differentiable. Its primal type is normally `Integer` or `Bool`.
```

#### `Δx`, `∂x`, `dx`
ChainRules uses these perhaps atyptically.

> As a notation that is the same across propagators, regardless of direction. (Incontrast see `ẋ` and `x̄` below)

Reviewer 2: ```suggestion
As a notation that is the same across propagators, regardless of direction (incontrast see `ẋ` and `x̄` below).
```
